### PR TITLE
Kubernetes Support Using Pluggable Schedulers

### PR DESCRIPTION
**Pull request by mccheah**
Builds on #81 with the Kubernetes support. Notice that the Kubernetes components have no coupling to CoarseGrainedSchedulerBackend or TaskSchedulerImpl.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -320,6 +323,10 @@ object SparkSubmit extends CommandLineUtils {
 
     // The following modes are not supported or applicable
     (clusterManager, deployMode) match {
+      case (KUBERNETES, CLIENT) =>
+        printErrorAndExit("Client mode is currently not supported for Kubernetes.")
+      case (KUBERNETES, CLUSTER) if args.isPython || args.isR =>
+        printErrorAndExit("Kubernetes does not currently support python or R applications.")
```
Comment (ash211) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: this shows an unresolved problem of the pluggability: how to make SparkSubmit pluggable in a way where adding this k8s executor provider doesn't require modifying SparkSubmit.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 2]
```diff
@@ -186,6 +192,18 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
       .getOrElse(sparkProperties.get("spark.executor.instances").orNull)
     keytab = Option(keytab).orElse(sparkProperties.get("spark.yarn.keytab")).orNull
     principal = Option(principal).orElse(sparkProperties.get("spark.yarn.principal")).orNull
+    kubernetesMaster = Option(kubernetesMaster)
```
Comment (ash211) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: same here -- wouldn't want any modifications to this SparkSubmitArguments class in the k8s addition

File: kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+import org.apache.spark.deploy.rest.kubernetes._
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils
+
+private[spark] class Client(
+    sparkConf: SparkConf,
+    mainClass: String,
+    mainAppResource: String,
+    appArgs: Array[String]) extends Logging {
+  import Client._
```
Comment (ash211) [inline kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: what's this importing?  from what `Client` ?

File: kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+  // TODO set precise version by default
+  private val driverDockerImage = sparkConf.get(
+    "spark.kubernetes.driver.docker.image", "spark-driver:latest")
+  private val uploadedDriverExtraClasspath = sparkConf
+    .getOption("spark.kubernetes.driver.uploads.driverExtraClasspath")
+  private val uploadedJars = sparkConf.getOption("spark.kubernetes.driver.uploads.jars")
+
+  private val secretBytes = new Array[Byte](128)
+  SECURE_RANDOM.nextBytes(secretBytes)
+  private val secretBase64String = Base64.encodeBase64String(secretBytes)
```
Comment (ash211) [inline kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: don't keep a reference to secretBytes, like this:

```
private val secretBase64String = {
    val secretBytes = new Array[Byte](128)
    SECURE_RANDOM.nextBytes(secretBytes)
    Base64.encodeBase64String(secretBytes)
}
```

File: kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+            .withSelector(selectors)
+            .withPorts(servicePorts.asJava)
+            .endSpec()
+          .done()
+        sparkConf.set("spark.kubernetes.driver.service.name", service.getMetadata.getName)
+        sparkConf.setIfMissing("spark.driver.port", DRIVER_PORT.toString)
+        sparkConf.setIfMissing("spark.blockmanager.port", BLOCKMANAGER_PORT.toString)
+        val submitRequest = buildSubmissionRequest()
+        val submitCompletedFuture = SettableFuture.create[Boolean]
+        val secretDirectory = s"/var/run/secrets/spark-submission/$kubernetesAppId"
```
Comment (ash211) [inline kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: is this path bundled into the k8s client libraries in a way where we don't have to also hardcode it?

File: kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+        val podWatcher = new Watcher[Pod] {
+          override def eventReceived(action: Action, t: Pod): Unit = {
+            if ((action == Action.ADDED || action == Action.MODIFIED)
```
Comment (ash211) [inline kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: matching else to log ignored events at DEBUG?  or is this too high volume?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 1 above)
Comment (mccheah) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Hm, right - could be done but this iteration of the thought experiment does not include this.

File: kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala (hunk 3 above)
Comment (mccheah) [inline kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: It's from the companion object.

File: kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala (hunk 5 above)
Comment (mccheah) [inline kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: This is where the file is mounted, so it's specified by the client.

Comment (mccheah) [inline kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: I think /`var/run/secrets` is standard though, not sure if we could validate this anywhere.

**Pull request closed without merging**